### DISCUSSION TITLE
Timeframe for Craft/Reka Integration ☺

### DISCUSSION THREAD

**Discussion by matmanna** [prevwong/craft.js#532] [Q&A] [answered]
Hello @prevwong, I am very excited by the prospects of using Reka with Craft and I am wondering if there is any rough timeframe of when you expect that to be possible (this year, next year, etc.). I don't need a fixed date, but just wanted to inquire and ask for an estimate so that I can decide whether to begin using reka in a project that may eventually require craft.js too. 😊

Thank you,

**Comment by prevwong** [comment, marked as answer]
Hey @joshpinto6 

I've already started working on an early prototype for this new version of Craft with Reka. 

As for timelines, I anticipate perhaps 2-3 months for an alpha version that's usable.

What I've right now: 

![](https://cdn.discordapp.com/attachments/665481348311547924/1108313831647756308/Screenshot_2023-05-17_at_4.43.15_PM.png)

The dnd stuff is being worked on right now, and need to figure out some easy-to-use abstractions like we have with Craft right now (ie: the `useEditor`)

**Reply by matmanna** [reply]
That looks so cool, I can't wait to test the demo out!! 🤩 (Sorry for the late response)

**Comment by paulwongx** [comment]
Hi @prevwong, thanks for your work on this library and rekajs. Similar to OP, was wondering what the timeline for the integration with craftjs was. This demo [here](https://reka.js.org/) looks like the completed version of what you were drafting up above, however, I don't see any dependecy to craftjs in the site `package.json`. 

Could you kindly help me understand the current status and any rough timelines for the integration into craftjs? 

Thanks again for your hard work on these libraries!

**Reply by matmanna** [reply]
That demo does not include craft.js, and does not support drag-and-drop or any other of the specific functionalities of Craft.js. The Craft.js integration demo that @prevwong showed is not available/released at this time.

**Comment by saifurrehmanx** [comment]
Hi @prevwong, could you please share the repository URL for Reka + Craft.js' codebase if it's publicly available? Thanks!

**Comment by Syed-Sheharyar** [comment]
According to the original answer here, it will take 2-3 months while now it is close to 6 months and there is still an alpha version not available. Please, make it public and get feedback. Welcome contributions, I am too curious to use it now.

**Reply by prevwong** [reply]
The original answer I gave here was an estimate and was never a guarantee plus it goes without saying - life gets in the way. I work on Craft because it’s something I enjoy doing and I believe it brings value to myself and to people who use it. More importantly, I do it because I want to and not because I owe you or anyone anything.

I indeed have been working on the integration ever since but I still don’t have something that I can share that would be ready for any meaningful use and collaboration.

If you want to help with this stage of the project, then feel free to use Reka by itself and provide feedback on it or better yet, consider sponsoring the project so I can dedicate more time to it. The integration with Craft will be shared when I have something more concrete.

**Reply by Syed-Sheharyar** [reply]
You said that use Reka yourself and add your drag and drop on top of it. I know Reka is only for State Management and is used to create React components with state and conditional rendering. Does Reka is highly granular which can work with my custom solution as well? I might try it with tiptap editor which supports React components. Users control components with Reka. Tiptap manages text editing and drag and drop. Tell me how's this idea?

**Reply by chrisedington** [reply]
Thanks @prevwong im very keen to see it and plan on sponsoring soon!

**Reply by chrisedington** [reply]
A question I have though is if I built a production app using Craft at the moment, and then at some point the Reka integration comes into play, how 'backward compatible' or 'forward compatible' do you estimate it to be?

Would it make sense to run at full speed now, or wait for the new integration?

I know it's a very broad question, but a general sense would be nice

**Reply by prevwong** [reply]
@chrisedington The biggest breaking change is the current Craft `EditorState` and the new Reka AST. For the most part, I think the current `EditorState` can be automatically translated into the new AST without any work from the developer's side. The only part of the current `EditorState` that might be a bit of an issue is the [Linked Nodes](https://craft.js.org/docs/concepts/user-components#defining-editable-elements), which I think it's not worth supporting moving forward since it has been difficult to maintain and causes a lot of confusion with users. 

Of course, I would still prefer if I could find a way to temporarily support this feature in the next version of Craft so no one will have any issues when migrating over to the new version, but I would first need to come up with a way to support Linked Nodes in the AST even if it's temporary. 

Apart from that, in terms of building page builder UIs - I am reusing a lot of the existing concepts and APIs from the current version so I don't think this should be an issue moving forward, plus I am planning an adapter for backwards compatibility so people could potentially incrementally switch over to the new APIs.

To summarise, if you can build a page editor with Craft today, then you should go ahead with it since chances are you don't need things like realtime collaboration, variables/expressions, repeated elements, conditionally rendering elements etc. Then when the new version is ready, your `EditorState` can be easily translated into the new AST. The only caveat is to perhaps avoid using things like Linked Nodes for now.

**Reply by chrisedington** [reply]
Thanks @prevwong for the detail :)
